packages/kaiju-mapper/src/kaiju_mapper/gmapper.py

```diff
@@ -0,0 +1,438 @@
+                ]  # Handle NaN
+
+                if sum(element_ad_scores) == 0:
+                    logger.info(
+                        f"Convergence after {iteration} iterations - all AD scores are zero."
+                    )
+                    return cover.intervals
+
+                found_valid = False
+                while not found_valid and len(all_elements_idx) > 0:
+                    # Sample one of the intervals weighted by ad score
+                    weights = np.asarray(element_ad_scores)[all_elements_idx]
+                    if weights.sum() == 0:
+                        # If all weights are zero, use uniform weights
+                        weights = np.ones_like(weights) / len(weights)
+                    else:
+                        weights = weights / weights.sum()
+
+                    current_element = int(
+                        np.random.choice(np.asarray(all_elements_idx), p=weights)
+                    )
+                    j = current_element
+
+                    if len(interval_membership[j]) == 0:
+                        removal_idx = all_elements_idx.index(j)
+                        all_elements_idx.pop(removal_idx)
+                        continue
+
+                    if self.ad_test(interval_membership[j]) > self.ad_threshold:
+                        check_interval[j] = True
+                        check_interval.insert(j + 1, True)
+
+                        new_intervals, interval_membership = self._split(
+                            interval_membership, cover.intervals, self.g_overlap, j
+                        )
+                        cover.intervals = new_intervals
+
+                        found_valid = True
+                    else:
+                        removal_idx = all_elements_idx.index(j)
+                        all_elements_idx.pop(removal_idx)
+
+                if not found_valid:
+                    logger.info(
+                        f"Convergence after {iteration} iterations - no valid splits found."
+                    )
+                    return cover.intervals
+
+            if len(cover.intervals) > self.max_intervals:
+                logger.info(
+                    f"Reached maximum number of intervals ({self.max_intervals})."
+                )
+                break
+
+        return cover.intervals
+
+    def ad_test(self, data):
+        """
+        Anderson-Darling Test for normality.
+
+        Parameters
+        ----------
+        data : list or np.ndarray
+            Data points to test for normality
+
+        Returns
+        -------
+        float
+            Anderson-Darling test statistic (corrected)
+        """
+        n = len(data)
+
+        if n == 0:
+            return 0
+        try:
+            and_corrected = anderson(data)[0] * (1 + 4 / n - 25 / (n**2))
+            return and_corrected
+        except Exception as e:
+            logger.warning(f"Anderson-Darling test failed: {e}")
+            return 0
```

Comment:
This method does not reference self, it probably should not be a method on this class. This happened a lot in this PR, 9425474a2ad951d97386ca1a9ffb74ade6276b44 shows the extraction of these methods to functions